common/src/stack/switch/pylib/switch/m7800.py

```diff
@@ -316,72 +318,256 @@ def del_partition_member(self, partition, guid):
 
+		If the toggle appears to fail, a SwitchException is raised.
+		"""
+		results = self.proc.ask(cmd = 'image boot next')
+		errors = self._get_errors(command_response = results)
+		if any(errors):
+			raise SwitchException(f'Setting next boot image failed with error {errors}')
 
 	def install_firmware(self, image):
-		self.proc.ask(f'image install {image}', timeout=1800)
+		"""Commands the switch to install the firmware image with the provided name.
 
+		The image must be previously loaded onto the switch with image_fetch().
+		If the install appears to fail, a SwitchException is raised.
+		"""
+		steps_regex = r'Step \d of (?P<max_steps>\d)'
+		results = self.proc.ask(cmd = f'image install {image}', timeout = 1800)
+		# try to get the number of steps expected to be performed
+		try:
+			num_steps = next(
+				(result for result in results if re.match(steps_regex, result, flags = re.IGNORECASE) is not None)
+			)
+		except StopIteration:
+			errors = self._get_expected_errors(command_response = results)
+			raise SwitchException(f'No firmware install steps appear to have been performed: {errors}')
+		# expect a number of success markers equal to the number of steps
+		num_steps = int(re.match(steps_regex, num_steps, flags = re.IGNORECASE).group('max_steps'))
+		completed_steps = len([result for result in results if '100.0%' in result])
+		if completed_steps != num_steps:
+			errors = self._get_expected_errors(command_response = results)
+			raise SwitchException(
+				f'Only {completed_steps} of {num_steps} firmware install steps appear to have completed successfully: {errors}'
+			)
 
 	def image_delete(self, image):
-		self.proc.say(f'image delete {image}')
+		"""Commands the switch to delete the firmware image with the provided name.
+
+		The image must have been previously loaded onto the switch with image_fetch().
+		If the deletion appears to fail, a SwitchException is raised.
+		"""
+		results = self.proc.ask(cmd = f'image delete {image}')
+		errors = self._get_errors(command_response = results)
+		if any(errors):
+			raise SwitchException(f'Image delete failed with error {errors}')
 
 
 	def image_fetch(self, url):
-		self.proc.ask(f'image fetch {url}', timeout=900)
+		"""Commands the switch to fetch a firmware image from the provided URL.
 
+		The URL must begin with one of the supported protocols or a SwitchException is raised.
+		If the transfer appears to fail, a SwitchException is raised.
+		"""
+		# validate the fetch url protocol is one we support
+		if not any((url.startswith(protocol) for protocol in self.SUPPORTED_IMAGE_FETCH_PROTOCOLS)):
+			raise SwitchException(f'Image fetch URL must be one of the following supported protocols {self.SUPPORTED_IMAGE_FETCH_PROTOCOLS}')
+
+		results = self.proc.ask(cmd = f'image fetch {url}', timeout = 900)
+		# check for success indicators and raise an error if not found.
+		if not any(('100.0%' in result for result in results)):
```

Comment:
> redundant set of parens. I believe there's a couple instances of this... list comps in any/next, too.

Thought I needed the extra set of parens to get the generator expression to work. I guess not in this context. Will remove.

> could this just be == instead of in? `in` is much more expensive, since it scans the whole string.

There's no guarantee that there won't be other characters in the string containing the `100.0%`, and until this is a demonstrable performance bottleneck I think this would be premature optimization.